elastica/rigidbody/mesh_rigid_body.py

```diff
@@ -0,0 +1,198 @@
+        self.distance_to_face_centers = _batch_norm(
+            self.face_centers - center_of_mass.reshape(3, 1)
+        )
+        self.direction_to_face_centers = (
+            self.face_centers - center_of_mass.reshape(3, 1)
+        ) / self.distance_to_face_centers
+        self.distance_to_faces = np.zeros((MaxDimension.value(), self.n_faces))
+        self.direction_to_faces = np.zeros(
+            (MaxDimension.value(), MaxDimension.value(), self.n_faces)
+        )
+        for i in range(MaxDimension.value()):
+            for k in range(self.n_faces):
+                self.distance_to_faces[i, k] = np.linalg.norm(
+                    self.faces[:, i, k] - center_of_mass
+                )
+                self.direction_to_faces[:, i, k] = (
+                    self.faces[:, i, k] - center_of_mass
+                ) / self.distance_to_faces[i, k]
+
+        self.mass_second_moment_of_inertia = mass_second_moment_of_inertia.reshape(
+            MaxDimension.value(), MaxDimension.value(), 1
+        )
+
+        self.inv_mass_second_moment_of_inertia = np.linalg.inv(
+            mass_second_moment_of_inertia
+        ).reshape(MaxDimension.value(), MaxDimension.value(), 1)
+
+        # position is at the center
+        self.position_collection = np.zeros((MaxDimension.value(), 1))
+        self.position_collection[:, 0] = center_of_mass
+
+        self.velocity_collection = np.zeros((MaxDimension.value(), 1))
+        self.omega_collection = np.zeros((MaxDimension.value(), 1))
+        self.acceleration_collection = 0.0 * self.velocity_collection
+        self.alpha_collection = 0.0 * self.omega_collection
+
+        self.director_collection = np.zeros(
+            (MaxDimension.value(), MaxDimension.value(), 1)
+        )
+        self.director_collection[0, ...] = normal
+        self.director_collection[1, ...] = binormal
+        self.director_collection[2, ...] = tangents
+
+        test_faces = np.zeros((3, 3, self.n_faces))  # dim,vertices,faces
+        for k in range(self.n_faces):
+            for m in range(3):
+                for i in range(3):
+                    test_faces[i, m, k] += center_of_mass[i]
+                    for j in range(3):
+                        test_faces[i, m, k] += (
+                            self.distance_to_faces[m, k]
+                            * self.director_collection[i, j, 0]
+                            * self.direction_to_faces[j, m, k]
+                        )
+
+        self.face_normals_lagrangian = np.zeros((MaxDimension.value(), self.n_faces))
+
+        # self.face_normals_lagrangian = self.director_collection[...,0].T@self.face_normals
+        for i in range(3):
+            for j in range(3):
+                for k in range(self.n_faces):
+                    self.face_normals_lagrangian[i, k] += (
+                        self.director_collection[j, i, 0] * self.face_normals[j, k]
+                    )
```

Comment:
I just realized by the same logic as above this can just be self.face_normals_lagrangian = self.face_normals.copy(), since the lab frame initial is the identity.